src/graphql/mod.rs

```diff
@@ -97,21 +98,69 @@ impl QueryClient {
         Q::ResponseData: Debug,
     {
         let body = Q::build_query(variables);
-        let request = self.inner().post(&self.api_url).json(&body);
 
         debug!("url: {:?}", &self.api_url);
-        debug!("request: {:#?}", request);
         debug!("GraphQL query: \n{}", body.query);
 
-        let response: Response<Q::ResponseData> = request.send().await?.json().await?;
+        let response: Result<Response<Q::ResponseData>, APIError> =
+            self.retry_request::<Q>(body, handler).await;
         debug!("GraphQL response: {:#?}", response);
 
-        if let Some(errors) = response.errors.clone() {
-            let first_error = errors[0].clone();
+        response
+    }
 
-            match check_auth_error(&first_error) {
-                Some(err) => return Err(err),
-                None => return handler(response, first_error),
+    // Attempts the request and retries up to 3 times if the request times out.
+    async fn retry_request<Q>(
+        &self,
+        body: QueryBody<Q::Variables>,
+        handler: impl Fn(
+            Response<Q::ResponseData>,
+            graphql_client::Error,
+        ) -> Result<Response<Q::ResponseData>, APIError>,
+    ) -> Result<Response<Q::ResponseData>, APIError>
+    where
+        Q: GraphQLQuery,
+        Q::ResponseData: Debug,
+    {
+        let mut retry_count = 0;
+        let request = self.inner().post(&self.api_url).json(&body);
+        let mut response: Response<<Q as GraphQLQuery>::ResponseData> =
+            request.send().await?.json().await?;
+
+        debug!("GraphQL response: {:#?}", response);
+
+        // We use <= 3 so it does one extra loop where the last response is checked
+        // in order to return an APIError::Timeout if it was a timeout error in the
+        // case of it failing all 3 retries.
+        while response.errors.is_some() && retry_count <= 3 {
+            if let Some(errors) = response.errors.clone() {
+                let first_error = errors[0].clone();
+
+                match check_retry_and_auth_error(&first_error) {
+                    Some(APIError::UnAuthenticated) => return Err(APIError::UnAuthenticated),
+                    Some(APIError::Timeout { domain }) => {
+                        if retry_count == 3 {
+                            return Err(APIError::Timeout { domain });
+                        }
+                        retry_count += 1;
+                        eprintln!(
+                            "... request to {domain} timed out, retrying the request {}/3",
+                            retry_count
+                        );
+                        debug!("request to {domain} timed out. Retry {retry_count}/3");
+
+                        thread::sleep(time::Duration::from_secs(5));
+
+                        let request = self.inner().post(&self.api_url).json(&body);
+
+                        debug!("request: {:?}", request);
```

Comment:
can you use the prettier debug format here? 

`debug!("request: {:#?}", request);`